src/server/frontend_wayland/wl_keyboard.cpp

```diff
@@ -94,27 +92,25 @@ void mf::WlKeyboard::event(MirKeyboardEvent const* event, WlSurface& surface)
     auto const timestamp = mir_input_event_get_wayland_timestamp(mir_keyboard_event_input_event(event));
     int const scancode = mir_keyboard_event_scan_code(event);
     uint32_t const wayland_state = down ? KeyState::pressed : KeyState::released;
-    send_key_event(serial, timestamp, scancode, wayland_state);
 
-    if (as_nullable_ptr(focused_surface) == &surface)
-    {
-        /*
-         * HACK! Maintain our own XKB state, so we can serialise it for
-         * wl_keyboard_send_modifiers
-         */
-        xkb_key_direction const xkb_state = down ? XKB_KEY_DOWN : XKB_KEY_UP;
-        xkb_state_update_key(state.get(), scancode + 8, xkb_state);
-
-        update_modifier_state();
-    }
-    else
+    if (focused_surface != surface)
```

Comment:
Um, how does this work? As far as I can tell `WlSurface` doesn't implement `operator==` or `operator!=`, so what is this comparing?

Reply:
`focused_surface` is a `Weak<>`, which has member function: `auto operator==(T const& other) const -> bool`.

Personally, I strongly prefer `operator==()` to be a non-member, and implement an equivalence relationship. (Neither of which is true for this case.)

Maybe the semantics of this `operator==()` are sufficiently confusing we should have objected when this was proposed (#1614).

But that's how this works.